### ISSUE TITLE
IPC issue on NetBSD 7.0.1 amd64

### ISSUE THREAD

**Issue by jibanes** [kevinlawler/kona#448]
Added a netbsd section in Makefile (copied from openbsd); ran gmake, screenshot attached.

![kona](https://cloud.githubusercontent.com/assets/1454666/17542019/0292faf0-5e7a-11e6-90ba-fb7e2ff4b3e7.png)

**Comment by jibanes** [comment]
Required changes are minimal; as such:

* netbsd ships with getline and getdelim, no need to add those.
* you can define getline_ to getline, getdelim_ to getdelim instead.

It _should_ work.

**Comment by jibanes** [comment]
I found a much greater issue on NetBSD, I don't know if other BSDs are affected, but please update this thread if you have the ability to test.

Kona Process 1:
k -i 5000

Kona Process 2:
  h: 3: `"127.0.0.1", 5000
5
  h 4: "b:1"
  h 4: "b:1"
1

The first assignment fails; the second has succeeded; it has succeeded because I pressed "enter" in the console of the Process 1. I haven't had the time to investigate further.

**Comment by tavmem** [comment]
Will try to get NetBSD working on one of my machines.
Changing the title of this issue to reflect the current problem.

**Comment by tavmem** [comment]
Just re-checked that it still works in Linux.   In process 1:
```
[tom@kona]$ rlwrap ./k -i 5000
K Console - Enter \ for help
```
in process 2:
```
[tom@kona]$ rlwrap ./k
K Console - Enter \ for help

  h: 3: `"127.0.0.1", 5000
3
  h 4: "b:1"
1
```
in process 1:
```
  \v
,`b
  b
1
```

**Comment by tavmem** [comment]
Status update:  There seems to be a problem with the NetBSD install process using a USB.  The bios on my Dell machine does not recognize the NetBSD install USB as a boot option when setting the boot sequence.  The Dell bios does recognize a Fedora install USB.  The problem has been reported to the NetBSD Project.

**Comment by tavmem** [comment]
It seems that the Dell box is expecting UEFI, which is not available on NetBSD-7.  There is the ability to override this preference and get BIOS by picking "legacy" mode, but first you need to disable "Secure Boot" against which the box issues multiple warnings.  
NetBSD-8 beta does provide UEFI, and NetBSD-support suggests that I try that.

**Comment by tavmem** [comment]
Tried to use "Legacy" mode on the Dell box.  It turns out that the "Legacy Option ROMs" need to be enabled for Legacy boot mode, and it is not clear how to do that.  The NetBSD-8 beta seems to be the better route.

**Comment by tavmem** [comment]
I've tried the NetBSD -8 beta.  I could not get the UEFI install option to work.  I have reported the problem to NetBSD-support.

**Comment by tavmem** [comment]
NetBSD-support steered me to a USB UEFI image that is recognized by the Dell setup as a bootable option.  Have not yet installed NetBSD-8, but it looks possible now.

**Comment by ibara** [comment]
This bug also affects OpenBSD/amd64 6.1-current.

**Comment by tavmem** [comment]
NetBSD-8.0_BETA-amd64-uefi-install.img
is recognized by the Dell setup as a valid option in the boot sequence.

However, when I choose it, I get the following error message:
"Operating System Loader has no signature.  Incompatible with SecureBoot.  All bootable devices failed Secure Boot verification."

This has been reported to NetBSD-support.

**Comment by bakul** [comment]
For the build case (the first comment) there is an easy fix as NetBSD support should be very similar to FreeBSD.  Just remove `|| defined(__NetBSD__)` from line `src/getline.c:95`.

**Comment by tavmem** [comment]
When I disable SecureBoot, the NetBSD-8.0 BETA install is accessible.

**Comment by tavmem** [comment]
Worked through all the questions posed by the "sysinst" of NetBSD-8.0-BETA.  The "sysinst" reports that NetBSD-8.0-BETA has been successfully installed on the Dell.  However, rebooting the system fails to start NetBSD-8.0-BETA.
  
Error message (upon reboot):
EXT4-fs error (device dm-0) ext4-iget:4762: inode#3145770: comm systemd-udevd:  bad extended attribute block 74955769315331

I am able to re-install Fedora-26 on the same machine from a USB with no problems at all.

Next steps:  Try re-install of NetBSD-8.0-BETA using various other options.  If that fails, attempt to activate legacy BIOS ROMs and attempt to install stable version NetBSD-7 using legacy mode.

**Comment by tavmem** [comment]
Upon further inspection, when rebooting (after the "successful install" of NetBSD-8.0-BETA), the system gets left in "emergency mode".

After logging in as root and executing "journalctl -xb"  it appears that there was an attempt start up Fedora 26, even though the complete disk was allocated to NetBSD-8.0-BETA by the sysinst:

Kernel command line :  BOOT_IMAGE=/vmlinuz-4.11.8-300.fc26.x86_64 root=/dev/mapper/fedora-root ro rd.lvm.lv=fedora/root rd.lvm.lv=fedora/swap rhgb quiet LANG=en_US.UTF-8

**Comment by tavmem** [comment]
Re-installed Fedora 26 on the Dell box.  It's probably better to use GNOME Boxes, and give up on attempting to do a native install of NetBSD.

**Comment by tavmem** [comment]
Attempted to install NetBSD-7.1-amd64-install.img on GNOME Boxes.  It failed.  Error messages:
```
Booting from Hard Disk...
NetBSD MBR boot
Error No operating system
No bootable device.
```
From documention on OSes supported by GNOME Boxes:
"You can use GNOME Boxes to install and run Windows, Ubuntu, Mint, openSUSE and many other Linux distributions in separate containers on one computer."

Looked into VirtualBox from Oracle.   The list of viable guest OSes does not include NetBSD.  See
https://www.virtualbox.org/wiki/Guest_OSes

Next Step:  Return to previous plan ... Try to activate legacy BIOS ROMs on Dell box.   Attempt to install NetBSD-7.1 using legacy methods.

**Comment by ibara** [comment]
Might I suggest trying OpenBSD with Virtualbox. It works, and almost certainly has the same solution as NetBSD.

**Comment by tavmem** [comment]
Yes, OpenBSD and FreeBSD seem to be viable options.

Significant work has been done to make NetBSD a supported guest OS under VirtualBox.
However, there are some missing features (about 3 months of work).  
See   http://wiki.netbsd.org/projects/project/virtual-box-guest-tools/

**Comment by ibara** [comment]
I would hesitate on recommending FreeBSD to track down this bug. For one, the bug hasn't been reproduced on FreeBSD. And also, OpenBSD and NetBSD are much more similar to each other than either is to FreeBSD so you might run across fundamental operating system differences if using FreeBSD (and, indeed, the bug may not even show on FreeBSD).

**Comment by tavmem** [comment]
FYI:
Status:   VirtualBox installs on Dell box under Fedora 26, but any guest OS (e.g., OpenBSD or NetBSD) fails to launch.
Next Step:  Continue to research fixes.

[tom@new-host-3 ~]$ sudo rcvboxdrv setup
vboxdrv.sh: Stopping VirtualBox services.
vboxdrv.sh: Building VirtualBox kernel modules.
vboxdrv.sh: Starting VirtualBox services.
vboxdrv.sh: Building VirtualBox kernel modules.
vboxdrv.sh: failed: modprobe vboxdrv failed. Please use 'dmesg' to find out why.

Full text of "dmesg" is attached:
[dmesg.txt](https://github.com/kevinlawler/kona/files/1170221/dmesg.txt)

errors that occur once:
[    4.705404] tpm_crb MSFT0101:00: can't request region for resource [mem 0xfed40080-0xfed40fff]
[    4.705477] tpm_crb: probe of MSFT0101:00 failed with error -16

[    5.062946] iwlwifi 0000:01:00.0: Direct firmware load for iwlwifi-7265D-28.ucode failed with error -2
[    5.067813] iwlwifi 0000:01:00.0: capa flags index 3 larger than supported by driver

[    5.992180] audit: type=1130 audit(1500864828.776:69): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=lvm2-pvscan@8:3 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    7.840589] hid-sensor-hub 001F:8086:22D8.0004: timeout waiting for response from ISHTP device

error that occurs 5 times:
[    5.170479] pcieport 0000:00:1c.0: PCIe Bus Error: severity=Corrected, type=Physical Layer, id=00e0(Receiver ID)
[    5.170533] pcieport 0000:00:1c.0:   device [8086:9d14] error status/mask=00000001/00002000
[    5.170573] pcieport 0000:00:1c.0:    [ 0] Receiver Error         (First)

**Comment by ibara** [comment]
This isn't a *BSD dmesg--it's a Linux dmesg.
It looks like some of your laptop's hardware isn't supported by Fedora 26, and that Linux is doing the wrong thing in other cases.

**Comment by tavmem** [comment]
Next steps:  
-  Scrap Fedora 26 as the base system for the Dell box.  
-  Re-install Windows-10 on the Dell box from the recovery USB.
-  Install VirtualBox on Windows.
-  Install OpenBSD on the Windows/VirtualBox and see if OpenBSD launches there.

**Comment by ibara** [comment]
That should work.
Use this ISO image for VirtualBox: http://ftp4.usa.openbsd.org/pub/OpenBSD/snapshots/amd64/cd61.iso and choose to install sets via http when it asks.
Once you're up and running, issue `pkg_add git` and you should be ready to go.

**Comment by tavmem** [comment]
Status:  The Windows restore fails.

Details:  Before installing Fedora-26, I created Windows Recovery media on 2 separate USB sticks.
NetBSD-7.1 doesn't support UEFI. To disable UEFI, you need to disable SecureBoot.  Did so.
You also need to disable PTT.  
I mistakenly cleared PTT, instead of turning it off.  Turning off PTT was still not enough to disable UEFI, so I switched focus to NetBSD-8.0-BETA, which does suport UEFI.   However, I think that clearing PTT is the reason Windows won't restore now. (UEFI, SecureBoot & PTT are all on.)
PTT stores "keys" (among other things).

Next steps:  The Geek Squad at Best Buy trying to restore Windows.  (Box was purchased there 7/3/17.)
They should have a status update within 24 hours.  Meanwhile, I can attempt to install VirtualBox & OpenBSD on a different Windows machine and/or on a MAC.

**Comment by tavmem** [comment]
Got VirtualBox running on a different Windows Box.
The OpenBSD install appears to work, and at completion I get the message:
"CONGRATULATIONS! Your OpenBSD install has successfully completed!
To boot the new system, enter 'reboot' at the command prompt."

Upon entering 'reboot' at the command prompt, the install system starts all over again.
It's OpenBSD-ground-hog-day.

Next Step:  Check out the VirtualBox documentation in more detail.

**Comment by ibara** [comment]
You need to remove the CD from VirtualBox, as VB boots from that first if it finds it.

**Comment by tavmem** [comment]
Thanks !
OpenBSD is working in VirtualBox.  Installed git.  Cloned kona.  Obviously need to modify Makefile.

Mouse pointer integration is not supported (which is no big deal).
Is there anyway to increase the working window size, or scroll up/down in the working window?
The current working window has 25 lines and about 80 characters per line.

**Comment by ibara** [comment]
You don't need to modify the Makefile. Just run `pkg_add gmake` and issue `gmake` instead of `make`.

The window is 80x24 because that's the size of a console :)
Run `startx` to get X running. Note that the default window manager is fvwm (with cwm being the other in-base window manager). There are plenty of other window managers in the package repo if you'd rather use something else.

I highly recommend running tmux in your terminal no matter what, as it will give you scrollback. tmux is part of the OpenBSD base system, so it's already installed.
(Scrolling in tmux works like this: press Ctrl+b then the [ key. Then you can use the up and down arrows and page up/page down to scroll. Press q when you're done scrolling to return to input mode.)

**Comment by tavmem** [comment]
FYI -- The "Geek Squad" at Best Buy confirms that they cannot restore my copy of Windows using any of their tools.  Having "cleared" the PTT, none of the restore or recovery options work anymore.
My only "restore" alternative is to buy another copy of Windows for $100.

Since I cannot install VirtualBox when using Fedora-26 as the base OS on that machince, I authorized them to proceed.

**Comment by bakul** [comment]
I installed NetBSD in a VirtualBox VM. Once I added a section for netbsd (cloned from the one for freebsd) in the Makefile, everything built cleanly and the same tests pass on FreeBSD and NetBSD. [Will try OpenBSD as well if I find time]

Now on to the IPC issue!

1. I started k -i 5000 on a FreeBSD machine and used NetBSD k as client. This seems to work fine for the above tests. As well as on FreeBSD.

2. I started k -i 5000 on a NetBSD machine and used the FreeBSD k client. This works fine for the above tests. Note that there are IPC related bugs in all versions. Will create separate bug reports for them. In fact we need to create a suite of tests for this case.

3. I next used both client and server on the same machine. This too works fine.

**Comment by tavmem** [comment]
I started 2 kona-sessions in OpenBSD on the same VirtualBox VM.  I was able to reproduce the problem (first amend for each variable results in null, subsequent amends are OK), and I found a couple of more problems (a crash, and a wrong message).  
However, I made no changes to the Makefile to add a section for OpenBSD ... just used gmake ... and everything built cleanly.  Is it possible that not adding a new section causes some of the problems?

in session 1:
```
  rlwrap ./k -i 5000
```
in session 2:
```
  h: 3: `"127.0.0.1", 5000
  h: 4: "a:1"
  h: 4: "b:2"
  h: 4: "c:3"
```
in session 1:
```
  \v
`a `b `c
  a
  b
  c
```
in session 2:
```
  h: 4: "a:11"
11
  h: 4: "b:22"
22
  h: 4: "c:33"
33
```
in session 1:
```
  \v
`a `b `c
  a
11
  b
22
  c
33
```
in session 2:
```
  3: h
```
in session 1:
```
k: warning: k crashed, killed by SIGSEGV (core dumped)
k itself has not crashed, but for transparency,
it will now kill itself with the same signal
```

The wrong message is "k itself has not crashed".   Should be "rlwrap has not crashed".

**Comment by tavmem** [comment]
There already is a section for OpenBSD in Makefile.

**Comment by bakul** [comment]
Installed openbsd 6.1. It works as well as freebsd, netbsd, macos in terms of tests passed and IPC.

Tom, note that your openBSD tests are wrong. The correct use is `h 4: "a:11"`, not `h: 4: "a:11"`. 

FYI: I am running
- FreeBSD-10.3 i386 and amd64 (native) -- needed one change beyond current HEAD (@3d0ca5a)
- OpenBSD-6.1 amd64 (VM) -- built from current HEAD
- NetBSD-7.1 amd64 (VM) -- needed one change (Makefile) beyond current HEAD
- Darwin-16.6.0 amd64 (native) -- built from current HEAD

Debugging hints:
- make sure you are running the latest github.com/kevinlawler/kona sources.
- in your bug report precisely describe any local changes you had to make.
- If possible, try to use the above versions (and on x86_64). If you don't see a problem here but do see it on i386 or some other processor arch and OS version, try to see what is different. I'll try to help you with suggestions etc but basically you're on your own.
- for as yet unreleased version of any OS you're definitely on your own!
- if you see a problem on x86_64 with the OS version I have, report the precise symptom and test so that I can try to reproduce the problem locally.

**Comment by tavmem** [comment]
Sorry about the typos.  The syntax for the OpenBSD tests was correct.  I could not get the "shared clipboard" to work between the guest copy of OpenBSD and the host copy of Windows, so I transcribed the sessions by hand.  I will make edits to the notes to remove the typos.

Thanks for you suggestions and help.

**Comment by tavmem** [comment]
I was able to recreate the problem on a Dell Laptop (Intel Pentium) running Windows 10 Home Edition.
Open BSD-6.1 amd64 was installed on VirtualBox from http://ftp4.usa.openbsd.org/pub/OpenBSD/snapshots/amd64/cd61.iso
Sets were installed via http.

**Comment by tavmem** [comment]
FYI:   Just an interesting "status"  ... The repository 
https://github.com/tavmem/konaStk
is a version of kona that prints out (as much as possible) every step of what kona is doing (kona with Stack reporting).

konaStk works fine in the problem test case, when run in 2 konaStk sessions on OpenBSD-6.1 on the VirtualBox hosted by Windows 10 Home Edition.

NB:  konaStk has its own problems:  It only works on Linux or BSD.
You can also note that if you repeatedly display a variable the reference count goes up each time.
(That may also happen in kona ... I haven't checked.)

Anyway ... konaStk is useful in determining where kona deviates in its behavior.

**Comment by ibara** [comment]
@bakul "for as yet unreleased version of any OS you're definitely on your own!"
Please don't say this unless you know for certain it is always true. Your statement is, for example, totally untrue on OpenBSD. I will ONLY consider this matter resolved for OpenBSD when it is fixed on -current (a "yet unreleased version") and -current is always supported, always. This is why I had @tavmem install OpenBSD from the snapshots directory instead of the 6.1 directory.

**Comment by bakul** [comment]
It's a question of limited and unpaid resources. You're more than welcome to find and fix the problem yourself and get changes merged upstream. Since things work for 6.1, at least you've a baseline from which you can debug.

**Comment by ibara** [comment]
@bakul what are you talking about? If you're trying to argue OpenBSD development with me, you should stop right now.

**Comment by bakul** [comment]
Just kona development!

**Comment by tavmem** [comment]
Some more details on the behavior (with Windows 10 host, OpenBSD guest, on VirtualBox):
- sends null on initial amend, same amend fails multiple times, even if hit "enter
- after changing the amend to a new value, it works consistently

In the 2nd session:
```
$ rlwrap k
  h: 3: `"127.0.0.1",5000
3
  h 4: "a:123"
  h 4: "a:123"
  h 4: "a:123"
                    / hit "enter"
  h 4: "a:123"
                    / hit "enter"
  h 4: "a:123"
  h 4: "a:1"        / works when amend value is changed, then works consistently
1
  h 4: "a:1"
1
                    / hit "enter"
  h 4: "a:1"
1
  h 4: "a:222"
222
  h 4;  "b:33"      / new variable works immediately
33
```

**Comment by tavmem** [comment]
The behavior listed just above is not repeatable.

**Comment by tavmem** [comment]
So far, it seems that OpenBSD handles sockets and/or the Connection Pool differently than the other OSes.
That is consistent with the observed sporadic behavior (sometimes it works, sometimes it doesn't).

**Comment by tavmem** [comment]
( 4: dyad ) not working properly in OpenBSD is a recurring regression.

By applying the fix described in the commit of Jul-20-2017 577eeb5a6aa430080a93706f2ac03c9a21c27e89 
titled: “OpenBSD and NetBSD have getline() and getdelim()”
to any commit when compiling on OpenBSD ...

( 4: dyad ) works as early as Nov-11-2011 in the commit 13957c0247eaf84a44fe0199675f90333484bb62 
titled “Check OSX version (use previous behavior for < Lion). “

It fails beginning in Jan-18-2013 when Thread-Local Storage is introduced in commit 27ca93b1ec3b97ec413dd8dbc6de09ea910feee1 
titled “make 3 global variables thread-local”.

It works fine again beginning in Apr-23-2015 with the commit 1008af4f4aece9df12a83bc927f8270c62c82bd0 
titled: “thread-local storage on OSX with GCC”

While it continues to function, on Jun-17-2015 “ct-D” begins to appear on the server side
when the client disconnects using ( 3: monad ) beginning with the commit 8b796e0c466311044c65f38ffb6a3c2ae0409c5e 
titled: “http-GET-request not implemented in Windows”

Things get worse, as we begin to get “domain error” when the client executes ( 3: dyad )
beginning in Dec-22-2015 with the commit 508b67cf975f51da416c211ece2674ab7c498168 
titled: “merge pahihu through Dec 22, 2015”

We begin to get the current (inconsistent) behavior beginning in Apr-18-2016 with the commit edf4df889b1a87bd9e6c1334ce15a1276e78eec5 
titled: “435: k vs. k_test”

Note that the current behavior appears (partly) to be a re-initialization problem.
You can consistently get the correct behavior (up to a point) by hitting the “enter” key before each new command on both the server and the client.   However, when the client disconnects, the server crashes with a segfault.
